### PR TITLE
point collection

### PR DESCRIPTION
**Pull request by ChristianBeilschmidt**

### CODE REVIEW

File: src/collections/feature_collection.rs [hunk 1]
```diff
@@ -41,7 +55,16 @@ mod test {
         fn is_simple(&self) -> bool {
             unimplemented!()
         }
-        fn data(&self, _field: &str) -> Result<FeatureDataRef> {
+        fn data(&self, _column: &str) -> Result<FeatureDataRef> {
+            unimplemented!()
```
Comment (jdroenner) [inline src/collections/feature_collection.rs:59]: can we add tests for this?/ how complex is it to build a FeatureCollection without Arrow?

File: src/collections/point_collection.rs [hunk 2]
```diff
@@ -81,12 +81,12 @@ impl PointCollection {
     pub fn empty() -> Self {
         Self {
             data: {
-                let fields = vec![
-                    Field::new(Self::FEATURE_FIELD, Self::multi_points_data_type(), false),
-                    Field::new(Self::TIME_FIELD, Self::time_data_type(), false),
+                let columns = vec![
+                    Field::new(Self::FEATURE_COLUMN, Self::multi_points_data_type(), false),
+                    Field::new(Self::TIME_COLUMN, Self::time_data_type(), false),
```
Comment (jdroenner) [inline src/collections/point_collection.rs]: How will we handle features without time? will each feature use [min, max) ?

File: src/collections/point_collection.rs [hunk 3]
```diff
@@ -119,15 +119,15 @@ impl PointCollection {
     /// assert_eq!(pc.len(), 2);
     /// ```
     pub fn from_data(
-        coordinates: Vec<Vec<Coordinate>>,
+        coordinates: Vec<Vec<Coordinate2D>>,
```
Comment (jdroenner) [inline src/collections/point_collection.rs:123]: Since this is a multi-point collection, i think we want to rename PointCollection to MultiPointCollection to reflect this.

File: src/collections/point_collection.rs [hunk 4]
```diff
@@ -81,12 +81,12 @@ impl PointCollection {
...
-                let fields = vec![
-                    Field::new(Self::FEATURE_FIELD, Self::multi_points_data_type(), false),
-                    Field::new(Self::TIME_FIELD, Self::time_data_type(), false),
+                let columns = vec![
+                    Field::new(Self::FEATURE_COLUMN, Self::multi_points_data_type(), false),
+                    Field::new(Self::TIME_COLUMN, Self::time_data_type(), false),
                 ];
 
-                StructArray::from(ArrayData::builder(DataType::Struct(fields)).len(0).build())
+                StructArray::from(ArrayData::builder(DataType::Struct(columns)).len(0).build())
```
Comment (jdroenner) [inline src/collections/point_collection.rs:90]: I guess it is ok to use "DataType" and "ArrayData" without scope or rename in this case. However, we have to be careful how we handle this in the project. There will be a (Gdal) DataType type and maybe more.

File: src/collections/feature_collection.rs [hunk 5]
```diff
@@ -15,17 +15,31 @@ pub trait FeatureCollection {
     fn is_simple(&self) -> bool;
 
     /// Reserved name for feature column
-    const FEATURE_FIELD: &'static str = "__features";
+    const FEATURE_COLUMN: &'static str = "__features";
 
     /// Reserved name for time column
-    const TIME_FIELD: &'static str = "__time";
+    const TIME_COLUMN: &'static str = "__time";
```
Comment (jdroenner) [inline src/collections/feature_collection.rs]: FEATURE_COLUMN_NAME ?

File: src/collections/feature_collection.rs [hunk 6]
```diff
@@ -1,4 +1,4 @@
-use crate::primitives::FeatureDataRef;
+use crate::primitives::{FeatureData, FeatureDataRef, TimeInterval};
 use crate::util::Result;
 
 /// This trait defines common features of all feature collections
```
Comment (jdroenner) [inline src/collections/feature_collection.rs:4]: The FeatureCollection does not provide information about its spatial and temporal bounds and the used projection. Since this is something we want to have for all spatio temporal data types we should add traits for this (maybe similar to raster).

File: src/collections/point_collection.rs [hunk 7]
```diff
@@ -119,15 +119,15 @@ impl PointCollection {
...
 
-        let mut fields = vec![
-            Field::new(Self::FEATURE_FIELD, Self::multi_points_data_type(), false),
-            Field::new(Self::TIME_FIELD, Self::time_data_type(), false),
+        let mut columns = vec![
+            Field::new(Self::FEATURE_COLUMN, Self::multi_points_data_type(), false),
+            Field::new(Self::TIME_COLUMN, Self::time_data_type(), false),
         ];
 
         let mut builders: Vec<Box<dyn ArrayBuilder>> = vec![
```
Comment (jdroenner) [inline src/collections/point_collection.rs:138]: Please explain how the builders are used here. Why cant we use `pub fn append_slice(&mut self, v: &[T::Native]) -> Result<()>` of `PrimitiveBuilder` for the coordinates?

File: src/collections/point_collection.rs [hunk 8]
```diff
@@ -165,22 +165,22 @@ impl PointCollection {
...
             builders.push(feature_data.arrow_builder()?);
 
             data_types.insert(name, FeatureDataType::from(&feature_data));
         }
 
-        let mut struct_builder = StructBuilder::new(fields, builders);
+        let mut struct_builder = StructBuilder::new(columns, builders);
         for _ in 0..capacity {
             struct_builder.append(true)?;
         }
```
Comment (jdroenner) [inline src/collections/point_collection.rs:191]: why is this required? Is there a more intuitive way to generate the structs?

File: src/collections/point_collection.rs [hunk 9]
```diff
@@ -335,69 +299,283 @@ impl FeatureCollection for PointCollection {
...
+    /// } else {
+    ///     unreachable!();
+    /// }
+    /// if let FeatureDataRef::Number(numbers) = extended_collection.data("bar").unwrap() {
+    ///     assert_eq!(numbers.data(), &[2., 4.]);
+    /// } else {
+    ///     unreachable!();
+    /// }
+    /// ```
+    fn extend(&self, new_column: &str, data: FeatureData) -> Result<Self> {
```
Comment (jdroenner) [inline src/collections/point_collection.rs]: add_column

File: src/collections/point_collection.rs [hunk 10]
```diff
@@ -474,10 +652,10 @@ fn coordinates_filter(features: &ListArray, filter_array: &BooleanArray) -> Resu
                 let old_floats_array = old_coordinates
```
Comment (jdroenner) [inline src/collections/point_collection.rs]: Could you also do zip(filter_array, coordinate_array)?

File: src/collections/point_collection.rs [hunk 11]
```diff
@@ -504,7 +682,7 @@ fn time_interval_filter(
         }
 
         let old_timestamps_ref = time_intervals.value(feature_index);
-        let old_timestamps: &Date64Array = old_timestamps_ref.as_any().downcast_ref().unwrap();
+        let old_timestamps: &Date64Array = old_timestamps_ref.as_any().downcast_ref().unwrap(); // must obey type
```
Comment (jdroenner) [inline src/collections/point_collection.rs]: this call is used many times in this file. can we put it into a function which clearly states what it does and inline it if that is needed?

File: src/primitives/coordinate.rs [hunk 12]
```diff
@@ -92,15 +92,15 @@ impl Into<(f64, f64)> for Coordinate {
...
 
-impl Into<[f64; 2]> for Coordinate {
+impl Into<[f64; 2]> for Coordinate2D {
     fn into(self) -> [f64; 2] {
         [self.x, self.y]
     }
 }
 
-impl<'c> Into<&'c [f64]> for &'c Coordinate {
+impl<'c> Into<&'c [f64]> for &'c Coordinate2D {
```
Comment (jdroenner) [inline src/primitives/coordinate.rs:101]: do we really need this? one could also do let x: [f64; 2] = (*x).into(). There will be no performance hit.

**Review by jdroenner — COMMENTED** [review]
I would like to have the basic types not tied to arrow. it would be great if we could exclude it from FeatureDataRef. i think there should be a way to do this. Also, i think the DataRef trait is identical to AsRef<T>, so we could also implement AsRef for it.

File: src/collections/point_collection.rs (hunk 2 above)
Comment (ChristianBeilschmidt) [reply src/collections/point_collection.rs]: I'm going to introduce a default timestamp from min to max. This way we do not have to implement special cases and it is always valid.

File: src/collections/feature_collection.rs (hunk 1 above)
Comment (ChristianBeilschmidt) [reply src/collections/feature_collection.rs:59]: It is not intended to implement a feature collection without Arrow. There will be one implementation.

There are tests in the implementations. I only created tests for functions that are implemented in the trait itself, e.g. `is_empty`.

File: src/collections/feature_collection.rs (hunk 6 above)
Comment (ChristianBeilschmidt) [reply src/collections/feature_collection.rs:4]: Yeah, we have to create a type for this and add this to the different types. I'm voting to put this into a separate task.

File: src/collections/point_collection.rs (hunk 4 above)
Comment (ChristianBeilschmidt) [reply src/collections/point_collection.rs:90]: I could use `arrow` as the namespace if you like. Otherwise we wait for conflicts and decide then.

File: src/collections/point_collection.rs (hunk 7 above)
Comment (ChristianBeilschmidt) [reply src/collections/point_collection.rs:138]: The multi-points and time intervals are not PrimitiveTypes, they are Lists and fixed sized lists. This way, for other applications, we have something like this: `<f64, f64>, ..., <f64, f64>` rather than arrays of binary data. Moreover, multi-points are variable-sized lists of these 2-tuples.

File: src/collections/point_collection.rs (hunk 8 above)
Comment (ChristianBeilschmidt) [reply src/collections/point_collection.rs:191]: I'm currently not seeing a good way of doing it differently when using the builder. We could switch to using the buffers directly, but I am a bit concerned that this is more brittle or error-prone. I guess we could do some benchmarks and then think of improving it. But for now, I do not see better available functions that facilitate this process stylistically.

File: src/primitives/coordinate.rs (hunk 12 above)
Comment (ChristianBeilschmidt) [reply src/primitives/coordinate.rs:101]: It is not about performance, but there are APIs requiring slices of `f64`s.

File: src/collections/point_collection.rs (hunk 10 above)
Comment (ChristianBeilschmidt) [reply src/collections/point_collection.rs]: Unfortunately, I can't. The coordinates are lists of fixed size lists of floats and the filter is only a primitive boolean array. Thus, they vary in size and can't be zipped that easily.

I hope that the arrow filter kernel will be able to filter complex arrays in the future. Then we can delete the whole method.

File: src/collections/point_collection.rs (hunk 3 above)
Comment (ChristianBeilschmidt) [reply src/collections/point_collection.rs:123]: 652f06b

File: src/collections/feature_collection.rs (hunk 5 above)
Comment (ChristianBeilschmidt) [reply src/collections/feature_collection.rs]: 44c5ad2

File: src/collections/point_collection.rs (hunk 9 above)
Comment (ChristianBeilschmidt) [reply src/collections/point_collection.rs]: 10d2cdb

File: src/collections/point_collection.rs (hunk 11 above)
Comment (ChristianBeilschmidt) [reply src/collections/point_collection.rs]: 4e1af07

**Comment by ChristianBeilschmidt** [comment]
I used AsRef instead of DataRef in f1ec534
